### PR TITLE
LPS-122467 Use the parameter SEGMENTS_ENTRY_IDS from the HttpServletRequest

### PR DESCRIPTION
**Pull request by cgoncas**
Hey @shuyangzhou and  @slnn!!

I think we can use the HTTPRequest attribute SEGMENTS_ENTRY_IDS (that we already filling) to avoid recalculate the segment entry Ids.

Also, we have noticed that if we don't reuse this parameter, an infinite look could be generated https://issues.liferay.com/browse/PTR-2021.

@slnn could you test this solution?

@shuyangzhou , what do you think?

Thanks in advance!!

### CODE REVIEW

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

File: modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java [hunk 1]
```diff
@@ -118,7 +118,14 @@ protected void doRun(
...
+		long[] segmentsEntryIds = (long[])httpServletRequest.getAttribute(
+			SegmentsWebKeys.SEGMENTS_ENTRY_IDS);
+
+		if (segmentsEntryIds != null) {
+			return;
+		}
+
+		segmentsEntryIds = _getSegmentsEntryIds(
 			httpServletRequest, themeDisplay.getScopeGroupId(),
 			themeDisplay.getUserId());
```
Comment (cgoncas) [inline modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java:131]: This parameter was already set in the request here, but we didn't reuse it between ServicePreAction and SegmentsRoleContributor

**Comment by cgoncas** [comment]
Hey @shuyangzhou!! Looks like I broke a test :(. I am going to try to fix it. Thanks!!!

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by CarlosBrichete** [comment]
Just started reviewing :)

**Comment by CarlosBrichete** [comment]
ci:test:tango

**Comment by CarlosBrichete** [comment]
ci:bundle

**Comment by CarlosBrichete** [comment]
ci:test:bundle

**Comment by cgoncas** [comment]
Hey @shuyangzhou and @slnn!!

After checking it with QA (@CarlosBrichete ) looks like the broken test is not directly related to this pull, so:

@slnn could you run the Segments performance test with this fix?

Thanks!

**Comment by slnn** [comment]
OK, got it, will update when I got the test result.

**Comment by slnn** [comment]
Hi @cgoncas @shuyangzhou 

Results please see Test 15 and Test 15-1 in https://docs.google.com/spreadsheets/d/163CDSA1SIyLgdcKCBJvSHe5PhEDWt7iyHdu9p-FMsos/edit?usp=sharing

Detailed test results zip and cpu tracing please see here:https://drive.google.com/drive/folders/135KKaAF-L8sMM1dhUoRm6VlqWQvmctsw?usp=sharing

By the way, from the thread dump, I can see so many http apr threads of tomcat just kept WAITING or TIMED_WAITING state.

Lily

**Comment by shuyangzhou** [comment]
@cgoncas this is making things worse, those thread @slnn mentioned are all choking on search.  Please analyze the result and profile to adjust.

**Comment by cgoncas** [comment]
Hey @shuyangzhou !!

Thanks a lot for the review, we have deactivated some components that were a very important cause for the regression (SegmentationRoleContributor) so we will try to measure this pull again in https://github.com/liferay-tango/liferay-portal/pull/563

Thanks!!!

**Pull request closed without merging**